File: CAShapelayer/CAShapelayer/ViewController.m
//
//  ViewController.m
//  CAShapelayer
//
//  Created by Hanguang on 4/22/15.
//  Copyright (c) 2015 Hanguang. All rights reserved.
//

#import "ViewController.h"
#import "LHShapelayerView.h"
#import "LHCircularLoaderView.h"
#import "LHProgressView.h"

@interface ViewController ()
@property (nonatomic, strong) LHShapelayerView *shapeLayerView;
@property (nonatomic, strong) LHCircularLoaderView *progressView;
@property (weak, nonatomic) IBOutlet UIImageView *imageView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
//    self.shapeLayerView = [[LHShapelayerView alloc] init];
//    _shapeLayerView.frame = CGRectMake(0, 0, 150, 150);
//    _shapeLayerView.center = self.view.center;
//    [self.view addSubview:_shapeLayerView];
//    self.maskView = [[UIView alloc] initWithFrame:self.view.bounds];
//    _maskView.backgroundColor = [UIColor whiteColor];
//    [self.view addSubview:_maskView];
    self.progressView = [[LHCircularLoaderView alloc] initWithFrame:CGRectZero];
    _progressView.frame = self.imageView.bounds;
    _progressView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    [self.imageView addSubview:_progressView];
    
    [self performSelector:@selector(increaseProgress) withObject:nil afterDelay:0.3f];
    
}


- (void)increaseProgress {
    _progressView.indicatorProgress += 0.1f;
    if(_progressView.indicatorProgress < 1.0f)
        [self performSelector:@selector(increaseProgress) withObject:nil afterDelay:0.3f];
    else
        [self performSelector:@selector(reveal) withObject:nil afterDelay:0.4f];
}

- (void)reveal {
    [_progressView reveal];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: iOSRACSwift/iOSRACSwift/ViewController.swift
//
//  ViewController.swift
//  iOSRACSwift
//
//  Created by Hanguang on 1/18/16.
//  Copyright © 2016 Hanguang. All rights reserved.
//

import UIKit
import ReactiveCocoa

class ViewController: UIViewController {
    
    @IBOutlet weak var nameTextField: UITextField!
    @IBOutlet weak var passwordTextField: UITextField!
    @IBAction func loginButton(sender: UIButton) {
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        nameTextField.rac_textSignal()
            .toSignalProducer()
            .skip(2)
            .filter { (text) -> Bool in
                let string = text as! String
                return string.characters.count > 3
            }.startWithNext { (text) -> () in
                print(text!)
        }
//        let searchStrings = nameTextField.rac_textSignal()
//            .toSignalProducer()
//            .map { text in text as! String }
//            .throttle(0.5, onScheduler: QueueScheduler.mainQueueScheduler)
//        
//        let searchResult = searchStrings
//            .flatMap(.Latest) { (query: String) -> SignalProducer<(NSData, NSURLResponse), NSError> in
//                let URLRequest = NSURLRequest(URL: NSURL(string: "http://www.baidu.com")!)
//                return NSURLSession.sharedSession()
//                    .rac_dataWithRequest(URLRequest)
//                    .retry(2)
//                    .flatMapError { error in
//                        print("Error: \(error)")
//                        return SignalProducer.empty
//                }
//            }
//            .map { (data, URLResponse) -> String in
//                let string = String(data: data, encoding: NSUTF8StringEncoding)!
//                return string
//            }
//            .observeOn(UIScheduler())
//        
//        searchResult.startWithNext { (results) -> () in
//            print("Search results: \(results)")
//        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
}

